crates/sui-types/src/move_package.rs

```diff
@@ -103,18 +110,66 @@ pub struct MovePackage {
     linkage_table: BTreeMap<ObjectID, UpgradeInfo>,
 }
 
-/// Rust representation of upgrade policy constants in `sui::package`.
-pub const UPGRADE_POLICY_COMPATIBLE: u8 = 0;
-pub const UPGRADE_POLICY_ADDITIVE: u8 = 128;
-pub const UPGRADE_POLICY_DEP_ONLY: u8 = 192;
-
-pub fn is_valid_package_upgrade_policy(policy: &u8) -> bool {
-    [
-        UPGRADE_POLICY_COMPATIBLE,
-        UPGRADE_POLICY_ADDITIVE,
-        UPGRADE_POLICY_DEP_ONLY,
-    ]
-    .contains(policy)
+// NB: do _not_ add `Serialize` or `Deserialize` to this enum. Convert to u8 first  or use the
+// associated constants before storing in any serialization setting.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy)]
+pub enum UpgradePolicy {
+    /// Rust representation of upgrade policy constants in `sui::package`.
+    Compatible = 0,
+    Additive = 128,
+    DepOnly = 192,
+}
+
+impl UpgradePolicy {
+    /// Convenience accessors to the upgrade policies as u8s.
+    pub const COMPATIBLE: u8 = Self::Compatible as u8;
+    pub const ADDITIVE: u8 = Self::Additive as u8;
+    pub const DEP_ONLY: u8 = Self::DepOnly as u8;
+
+    pub fn is_valid_policy(policy: &u8) -> bool {
+        Self::try_from(*policy).is_ok()
+    }
+
+    pub fn check_compatibility(
+        &self,
+        old_module: &normalized::Module,
+        new_module: &normalized::Module,
+    ) -> PartialVMResult<()> {
+        match self {
+            Self::Compatible => {
+                let check_struct_and_pub_function_linking = true;
+                let check_struct_layout = true;
+                let check_friend_linking = false;
+                Compatibility::new(
+                    check_struct_and_pub_function_linking,
+                    check_struct_layout,
+                    check_friend_linking,
+                )
+                .check(old_module, new_module)
+            }
+            Self::Additive => InclusionCheck::Subset.check(old_module, new_module),
+            Self::DepOnly => InclusionCheck::Equal.check(old_module, new_module),
+        }
+    }
+}
+
+impl TryFrom<u8> for UpgradePolicy {
+    type Error = ();
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            x if x == Self::Compatible as u8 => Ok(Self::Compatible),
+            x if x == Self::Additive as u8 => Ok(Self::Additive),
+            x if x == Self::DepOnly as u8 => Ok(Self::DepOnly),
+            _ => Err(()),
+        }
+    }
+}
+
+impl Display for UpgradePolicy {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        (*self as u8).fmt(f)
+    }
 }
```

Comment:
Wouldn't it be fine (or better, even) to use the derived `Display` impl which displays it using the enumerand name?

Reply:
Totally fine to display the enumerated name, and you're probably right that it's better to do that. I was thinking of wanting the display to match the constant on-chain, but that doesn't really make much sense now that I'm thinking about it...